[WebNN EP] Support Trilu op

peishenyan opened the pull request:
Adds support for Trilu via WebNN Triangular op

Honry commented:
Please add this new op to `op_map` in `helper.h`, update the unit tests under `/js/web/test/suite-test-list.jsonc`, also the  `js/web/docs/webnn-operators.md`

Honry commented on onnxruntime/core/providers/webnn/builders/impl/triangular_op_builder.cc:
```diff
@@ -0,0 +1,103 @@
+    std::vector<uint8_t> unpacked_tensor;
+    ORT_RETURN_IF_ERROR(onnxruntime::utils::UnpackInitializerData(diagonal_tensor, unpacked_tensor));
+    const auto diagonal = *reinterpret_cast<int64_t*>(unpacked_tensor.data());
+    options.set("diagonal", static_cast<long>(diagonal));
```
Prefer using `gsl::narrow<int32_t>`

Honry reviewed (approved):
LGTM, thanks!

peishenyan commented:
Thanks for the comments. I have added the op map in `helper.h`, updated the `webn-operators.md`, and addressed the comments.

PTAL, @Honry @fdwr @guschmue  Thanks!

guschmue commented:
/azp run ONNX Runtime Web CI Pipeline,Windows GPU CI Pipeline, Linux Android Emulator QNN CI Pipeline

guschmue commented:
/azp run Linux CPU CI Pipeline,Linux CPU Minimal Build E2E CI Pipeline,Linux GPU CI Pipeline,Linux GPU TensorRT CI Pipeline,Linux OpenVINO CI Pipeline,Linux QNN CI Pipeline,MacOS CI Pipeline,Windows ARM64 QNN CI Pipeline,Windows CPU CI Pipeline

guschmue commented:
/azp run Windows GPU TensorRT CI Pipeline,onnxruntime-binary-size-checks-ci-pipeline,orttraining-linux-ci-pipeline,orttraining-linux-gpu-ci-pipeline,orttraining-ortmodule-distributed,Windows x64 QNN CI Pipeline,Big Models

fdwr commented on onnxruntime/core/providers/webnn/builders/impl/triangular_op_builder.cc:
```diff
@@ -0,0 +1,103 @@
+    return false;
+  const auto input_size = input_shape.size();
+  if (input_size < 2) {
+    LOGS(logger, VERBOSE) << "Triangular only support input size >= 2d shape, input is "
```
`2D`

fdwr replied:
`Triangular only supports input`

fdwr commented on onnxruntime/core/providers/webnn/builders/impl/triangular_op_builder.cc:
```diff
@@ -0,0 +1,103 @@
+
+  if (!GetTensorName(input_defs, 1).empty()) {
+    // Optional input diagonal is provided, use diagonal initializer data.
+    const auto diagonal_tensor = *initializers.at(input_defs[1]->Name());
```
(minor) On line 85 you get the name and then reuse it. Could have done that here too rather than query the name twice.

fdwr reviewed (commented):
Already submitted while I was OOF, but some small thoughts for future polish, else LGTM 👍. Thanks for adding triangular matrix.
